### PR TITLE
Refactor Semantic-MathML and Semantic-Collapse

### PR DESCRIPTION
**Pull request by dpvc**
This moves the MakeMML call from Semantic-Collapse to Semantic-MathML so that the enrichment is done completely by Semantic-MathML.  The rest of Semantic-Collapse is broken into two parts:  Semantic-Complexity, which does the complexity computations and adds the mactions based on that, and Semantic-Collapse, which does the automatic collapsing based on container width.  All three actions can be enabled and disabled separately (though the latter ones rely on the former ones, so disabling an earlier phase effectively disables the later ones).

I've updated the lab pages to configure the new features, and cleaned up some of the code there (in particular, the getting and setting of the options).

### CODE REVIEW

**Comment by dpvc** [comment]
I'll have to fix the conflicts with the accessibility_menu PR that I just merged.

**Comment by dpvc** [comment]
OK, it should be ready to merge, now.

**Comment by zorkow** [comment]
I think I just screwed that up by merging PR #75.

File: extensions/Semantic-Complexity.js [hunk 1]
```diff
@@ -0,0 +1,625 @@
+//
+//  A filter to convert the enhanced MathML to MathJax internal format
+//  so we can display it, while adding maction elements for parts that
+//  can be collapsed.  We determine this based on a "complexity" value
+//  and collapse those terms that exceed a given complexity.
+//
```
Comment (zorkow) [inline extensions/Semantic-Complexity.js]: Outdated comment.

File: extensions/Semantic-Complexity.js [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+      //  Add the filter into the post-input hooks (priority 100, so other
+      //  hooks run first, in particular, the enrichment hook).
+      //
+      HUB.postInputHooks.Add(["Filter",Complexity],100);
+    },
+    
+    //
+    //  The main filter (convert the enriched MathML to the
+    //  MathJax internal format, with collapsing).
+    //
```
Comment (zorkow) [inline extensions/Semantic-Complexity.js]: Update comment.

File: extensions/Semantic-Complexity.js [hunk 3]
```diff
@@ -0,0 +1,625 @@
...
+        attrNames:["id","actiontype","complexity","selection"], attr:{}, selection:2
+      });
+      if (mml.type === "math") {
+        var mrow = MML.mrow().With({
+          data: mml.data,
+          complexity: mml.complexity,
+          attrNames: [], attr: {}
+        });
+        mrow.attrNames.push("complexity");
+        for (var i = mml.attrNames.length-1; i >= 0; i--) {
```
Comment (zorkow) [inline extensions/Semantic-Complexity.js]: Why not simply 
for (var i = mml.attrNames.length-1, name; name = mml.attrName[i]; i--) {

attrNames should not contain an empty string.

File: extensions/Semantic-Complexity.js [hunk 4]
```diff
@@ -0,0 +1,625 @@
...
+    },
+
+    //
+    //  If a parent is going to be collapsible, if can call this
+    //  to put back a collapsed child (rather than have too many 
+    //  nested collapsings)
+    //
+    UncollapseChild: function (mml,n,m) {
+      if (m == null) m = 1;
+      if (mml.attr["data-semantic-children"].split(/,/).length === m) {
```
Comment (zorkow) [inline extensions/Semantic-Complexity.js]: These mml.attr["data-semantic-children"].split(/,/) occur often enough to warrant their own little utility function. SplitAttribute(mml, 'children') or similar.

File: extensions/Semantic-Complexity.js [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+      }
+      return mml;
+    },
+
+    //
+    //  For bigops, get the character to use from the largeop at its core.
+    //
+    Collapse_bigop: function (mml) {
+      if (mml.complexity > this.COLLAPSE.bigop || mml.data[0].type !== "mo") {
+        var id = mml.attr["data-semantic-content"].split(/,/); id = id[id.length-1];
```
Comment (zorkow) [inline extensions/Semantic-Complexity.js]: Why not just add a .pop() here?

File: extensions/Semantic-Complexity.js [hunk 6]
```diff
@@ -0,0 +1,625 @@
...
+      if (mml.complexity > this.COLLAPSE.subscript)
+        mml = this.MakeAction(this.Marker(this.MARKER.subscript),mml);
+      return mml;
+    },
+    Collapse_subsup: function (mml) {
+      this.UncollapseChild(mml,0,3);
+      if (mml.complexity > this.COLLAPSE.subsup)
+        mml = this.MakeAction(this.Marker(this.MARKER.subsup),mml);
+      return mml;
+    }
```
Comment (zorkow) [inline extensions/Semantic-Complexity.js]: I feel all these little functions are similar enough to eventually refactor into a factory class. But that's for another day.

File: extensions/Semantic-MathML.js [hunk 7]
```diff
@@ -22,96 +22,133 @@ MathJax.Extension.SemanticMathML = {
...
+  //  If we are not disabled,
+  //    Get the enriched MathML and parse it into the jax root.
+  //    Mark the jax as enriched.
+  //
   Filter: function (jax,id,script) {
     delete jax.enriched;
-    this.running = true;
-    if (!this.config.disabled) jax.enriched = sre.Enrich.semanticMathmlSync(jax.root.toMathML());
-    this.running = false;
+    if (!this.config.disabled) {
```
Comment (zorkow) [inline extensions/Semantic-MathML.js]: Change to if(this.config.disabled) return;

File: Semantics-Lab/Lab.js [hunk 8]
```diff
@@ -181,6 +214,18 @@ var Lab = {
...
   },
   
+  
+  //
+  //  Get DOM elements needed by the Lab
+  //
+  GetDomElements: function () {
+    var ids = ["input","container","enhance","overflow","renderer",
+               "mathml","example","container","width","range_output"];
+    for (var i = 0; i < ids.length; i++) {
```
Comment (zorkow) [inline Semantics-Lab/Lab.js]: Why not 
for (var i = 0, id; id = ids[i]; i++) {
?
Reduces two array accesses to one.

**Comment by zorkow** [comment]
The Struik example is now not working anymore. I assume that has to do with the option
  SemanticCollapse: {autoCollapse: true},
no longer exists.
Btw., in TeX.html and MathML.html is a misspelled option:
  AssistiveMML: {diabled: true},
That never seemed ot have made a difference...

**Comment by dpvc** [comment]
> in `TeX.html` and `MathML.html` is a misspelled option:
`AssistiveMML: {diabled: true},`
That never seemed ot have made a difference...

Thanks.  I'll fix it.  Turning off AssistiveMML was just because it was not needed for theses tests.  Leaving it enabled isn't a problem (just takes a bit longer to process).

**Comment by dpvc** [comment]
The Struik example works for me (I did remove the unwanted configuration, though that should not have prevented it from working).  Can you test again with the updates to see if you still have trouble?  Remember to clear cache first.

**Comment by dpvc** [comment]
I have made the other changes you suggested, so I'm ready for another review, or a merge.

**Comment by zorkow** [comment]
Removed two merge conflicts. O/w lgtm.

**Comment by dpvc** [comment]
Thanks.

**Pull request merged**